File: api/src/routes/bot.rs
use std::sync::RwLock;

use actix_web::{get, web, HttpResponse};
use sqlx::{Pool, Postgres};

use crate::bot::manager;
use crate::bot::types::BotExternalAction;
use crate::errors::*;
use crate::extractors::user_from_cookie::UserFromCookie;
use crate::models;
use crate::states::app_config::AppConfig;
use crate::twitch::id_api::validate_user_token;

#[get("/info")]
pub async fn get_bot_info(
    user: UserFromCookie,
    pool: web::Data<Pool<Postgres>>,
    app_config: web::Data<AppConfig>,
    bot: web::Data<RwLock<manager::BotManager>>,
) -> Result<HttpResponse, AppError> {
    let bot = bot.read()?;

    let status = bot.status()?;
    let channel_registry = bot.channel_registry.clone();

    let name = &app_config.chat_bot_username;
    let connected = matches!(status, manager::BotStatus::Connected(_));

    let mb_credentials = models::BotCredentials::get_by_username(&pool, name).await?;

    let credentials_state = match mb_credentials {
        Some(credentials) => {
            let token_is_valid = validate_user_token(&credentials.access_token).await?;

            if token_is_valid {
                models::CredentialsState::Valid
            } else {
                models::CredentialsState::Invalid
            }
        }
        None => models::CredentialsState::NotFound,
    };

    let connected_to_chat = channel_registry.get(&user.logged.username).is_some();

    Ok(HttpResponse::Ok().json(models::BotInfo {
        name: name.to_string(),
        credentials_state,
        connected,
        connected_to_chat,
    }))
}

#[get("/join")]
pub async fn join_chat(
    user: UserFromCookie,
    bot: web::Data<RwLock<manager::BotManager>>,
) -> Result<HttpResponse, AppError> {
    let bot = bot.read()?;
    let status = bot.status()?.clone();

    match status {
        manager::BotStatus::Connected(sender) => {
            sender
                .send(BotExternalAction::Join {
                    channel_name: user.logged.username.clone(),
                    user_id: Some(user.logged.id.clone()),
                })
                .await?;

            // TODO: Remove this, set this for testing purposes
            sender
                .send(BotExternalAction::Join {
                    channel_name: String::from("namelessw05"),
                    user_id: None,
                })
                .await?;

            Ok(HttpResponse::Ok().finish())
        }
        _ => Err(AppError::from(AppErrorType::BotDisconnected)),
    }
}

#[get("/leave")]
pub async fn leave_chat(
    user: UserFromCookie,
    bot: web::Data<RwLock<manager::BotManager>>,
) -> Result<HttpResponse, AppError> {
    let bot = bot.read()?;
    let status = bot.status()?.clone();

    match status {
        manager::BotStatus::Connected(sender) => {
            sender
                .send(BotExternalAction::Leave(user.logged.username.clone()))
                .await?;

            Ok(HttpResponse::Ok().finish())
        }
        _ => Err(AppError::from(AppErrorType::BotDisconnected)),
    }
}

#[get("/connect")]
pub async fn connect(
    user: UserFromCookie,
    app_config: web::Data<AppConfig>,
    pool: web::Data<Pool<Postgres>>,
    bot_manager: web::Data<RwLock<manager::BotManager>>,
) -> Result<HttpResponse, AppError> {

    if user.logged.username != app_config.chat_bot_username {
        Err(AppError::from(AppErrorType::Unauthorized))
    } else {
        let mut manager = bot_manager.write()?;
        let status = manager.status()?;

        match status {
            manager::BotStatus::Connected(_) => Ok(HttpResponse::Ok().finish()),
            manager::BotStatus::Disconnected => {
                models::BotCredentials::get_or_create(&pool, &models::CreateBotCredentials {
                    access: &user.session.access_token,
                    refresh: &user.session.refresh_token,
                    user_id: &user.logged.id,
                }).await?;

                manager.connect().await?;
                Ok(HttpResponse::Ok().finish())
            }
        }
    }
}

File: app/src/api/bot.ts
import { z } from 'zod'

import { PUBLIC_API_URL } from '$env/static/public'

import { handleHttpError } from './utils'

const botInfoApiValidator = z.object({
  name: z.string(),
  connected: z.boolean(),
  connectedToChat: z.boolean(),
  credentialsState: z.enum(['valid', 'invalid', 'notFound']),
})

type BotInfoApi = z.infer<typeof botInfoApiValidator>

export async function getBotInfo(fetchFn = fetch): Promise<BotInfoApi> {
  const response = await fetchFn(`${PUBLIC_API_URL}/bot/info`, {
    credentials: 'include'
  })

  if (!response.ok) {
    handleHttpError(response)
  }

  const json = await response.json()
  return await botInfoApiValidator.parseAsync(json)
}

export async function askBotToJoinChat(): Promise<void> {
  const response = await fetch(`${PUBLIC_API_URL}/bot/join`, {
    credentials: 'include',
  })

  if (!response.ok) {
    handleHttpError(response)
  }
}

export async function askBotToLeaveChat(): Promise<void> {
  const response = await fetch(`${PUBLIC_API_URL}/bot/leave`, {
    credentials: 'include',
  })

  if (!response.ok) {
    handleHttpError(response)
  }
}

export async function connectBotToTwitch(): Promise<void> {
  const response = await fetch(`${PUBLIC_API_URL}/bot/connect`, {
    credentials: 'include',
  })

  if (!response.ok) {
    handleHttpError(response)
  }
}
